### PR TITLE
Type all remaining tests

### PR DESCRIPTION
**Pull request by jakkdl**
Decided to do a speedrun through all the remaining test files, with the help of https://github.com/JelleZijlstra/autotyping

If you see anything incorrect, you're very welcome to edit it straight away - I didn't put a lot of effort on average into the type hints, so there's probably a decent amount of them that are bad in some way or another.

### CODE REVIEW

File: trio/testing/_fake_net.py
Comment (TeamSpen210) [inline trio/testing/_fake_net.py:27]: We probably want to guard this import, since `trio.testing` is public.

File: trio/_ssl.py
Comment (TeamSpen210) [inline trio/_ssl.py:243]: I think this does need to be invariant. For `SSLStream` it could be covariant since we never use it as a return value, but `SSLListener` does return it. Could be covariant for the former and invariant for the latter, but that's just confusing.

File: trio/_tests/test_threads.py
Comment (TeamSpen210) [inline trio/_tests/test_threads.py:465]: No need to guard this, they'll just set `__annotations__`?

File: trio/_tests/test_threads.py
Comment (TeamSpen210) [inline trio/_tests/test_threads.py:560]: Obviously not something to do here, but I'd think it'd make most sense to do a protocol, with only the 6 `acquire_*`/`release_*` methods. That's all that a task needs to be limited, the others are only relevant for `CapacityLimiter`'s simple token strategy. The implementations of `acquire()`/`release()` could be included in the protocol since they defer to the `on_behalf_of` methods.

**Comment by CoolCat467** [comment]
How do we fix the following errors `test_exports` raises?

>   jedi can't see the following symbols in trio.lowlevel:
  {'trio.lowlevel.RunVarToken': {'extra': {'previous_value', 'redeemed'},
                                 'missing': set()}}

>  mypy can't see the following symbols in trio.lowlevel:
  {'trio.lowlevel.RunVarToken': {'extra': {'previous_value', 'redeemed'},
                                 'missing': set()}}

File: trio/_core/_tests/test_guest_mode.py
Comment (CoolCat467) [inline trio/_core/_tests/test_guest_mode.py:547]: Any reason not to use `object` instead of `Any`?

File: trio/_core/_tests/test_instrumentation.py
Comment (CoolCat467) [inline trio/_core/_tests/test_instrumentation.py:34]: Doesn't matter a whole lot, but I think `Iterable` is best used in return type if it's returning an object that defines an `__iter__` method. Even though it's more text, I think `Generator` would be more applicable here.

File: trio/_core/_tests/test_ki.py
Comment (CoolCat467) [inline trio/_core/_tests/test_ki.py:7]: I would import from `collections.abc` instead

File: trio/_core/_tests/test_multierror.py
Comment (CoolCat467) [inline trio/_core/_tests/test_multierror.py:401]: Why old generic if we are using `__future__` annotations?

File: trio/_tests/test_highlevel_ssl_helpers.py
Comment (CoolCat467) [inline trio/_tests/test_highlevel_ssl_helpers.py:65]: `object`?

File: trio/_tests/test_subprocess.py
Comment (CoolCat467) [inline trio/_tests/test_subprocess.py:518]: Isn't `sig` a signal type object and `frame` a `FrameType` (from `types`) object?

File: trio/_tests/test_threads.py
Comment (CoolCat467) [reply trio/_tests/test_threads.py:465]: That should be correct yes, as far as I am aware.

File: trio/_tests/test_threads.py
Comment (CoolCat467) [inline trio/_tests/test_threads.py:595]: Unrelated to the types side of things but noticed this, could we change `canna` -> `cannot`

File: trio/_tests/test_timeouts.py
Comment (CoolCat467) [inline trio/_tests/test_timeouts.py:2]: `collections.abc` for everything but TypeVar

File: trio/_tests/test_tracing.py
Comment (CoolCat467) [inline trio/_tests/test_tracing.py:1]: `collections.abc`

File: trio/testing/_fake_net.py
Comment (CoolCat467) [inline trio/testing/_fake_net.py:24]: `collections.abc` for a lot of these

**Review by CoolCat467 — CHANGES_REQUESTED** [review]
Looks pretty good except a few things here and there.

**Comment by CoolCat467** [comment]
All of the `collections.abc` imports I noted would have been able to be fixed automatically if #2809 would go through, with ruff's rule UP035 (https://docs.astral.sh/ruff/rules/deprecated-import/)

File: trio/_core/_tests/test_instrumentation.py
Comment (TeamSpen210) [reply trio/_core/_tests/test_instrumentation.py:34]: I prefer to generally use `Iterator`, since it leaves you open to changing the returned object to something else in the future. The fact it's a generator is an implementation detail. Unless you do want the users to use generator-specific methods.

**Comment by jakkdl** [comment]
yeah I intentionally don't bother in large part because I expect it to get autofixed, but also doesn't really matter.

File: trio/_core/_tests/test_guest_mode.py
Comment (jakkdl) [reply trio/_core/_tests/test_guest_mode.py:547]: I'm pretty sure you'd get a typing error on line 516/551 in the `old_get_events(*args)` call. The high-effort solution (here and elsewhere in this PR) would be to copy the signature of it and not use `*args` at all.

File: trio/_core/_tests/test_ki.py
Comment (jakkdl) [reply trio/_core/_tests/test_ki.py:7]: going to leave that for some autofixer / I don't care enough to bother fixing it manually.

File: trio/_core/_tests/test_multierror.py
Comment (jakkdl) [reply trio/_core/_tests/test_multierror.py:401]: uh, no reason at all. Might've been something `pyannotate` did

File: trio/_tests/test_highlevel_ssl_helpers.py
Comment (jakkdl) [reply trio/_tests/test_highlevel_ssl_helpers.py:65]: it might raise a type error about the signature not matching the superclass, not sure. But with an empty body it doesn't matter either way.

File: trio/_tests/test_subprocess.py
Comment (jakkdl) [reply trio/_tests/test_subprocess.py:518]: :+1:

File: trio/_tests/test_threads.py
Comment (jakkdl) [reply trio/_tests/test_threads.py:595]: lol, I suspect it might be an intentional joke - objectively the error message is pretty unhelpful. I vote to keep it :pirate_flag:

File: trio/_core/_tests/test_multierror.py
Comment (jakkdl) [reply trio/_core/_tests/test_multierror.py:381]: yeah, I did it intentionally to mirror the function body. I realize now the proper way to do it would maybe be `(exc: T_Exception) -> T_Exception | RunTimeError`, but that's kinda overkill.

File: trio/testing/_fake_net.py
Comment (jakkdl) [reply trio/testing/_fake_net.py:27]: hmm, this should ideally be picked up by the `no test-requirements` CI run. Will look at it.

File: trio/_tests/test_threads.py
Comment (jakkdl) [reply trio/_tests/test_threads.py:560]: Created #2824

**Comment by jakkdl** [comment]
> How do we fix the following errors `test_exports` raises?
> 
> > jedi can't see the following symbols in trio.lowlevel:
> > {'trio.lowlevel.RunVarToken': {'extra': {'previous_value', 'redeemed'},
> > 'missing': set()}}
> 
> > mypy can't see the following symbols in trio.lowlevel:
> > {'trio.lowlevel.RunVarToken': {'extra': {'previous_value', 'redeemed'},
> > 'missing': set()}}

(I/Somebody should update this error message to be more helpful)
This is because `dir()` doesn't see the attributes in it. I don't really know why, but setting `slots=True` seems to fix it and I don't think there's any particular reason not to.

File: trio/_tests/test_file_io.py
Comment (jakkdl) [reply trio/_tests/test_file_io.py:249]: thanks, that worked!

File: trio/_tests/test_threads.py
Comment (jakkdl) [reply trio/_tests/test_threads.py:465]: correct, thanks~

File: trio/testing/_fake_net.py
Comment (jakkdl) [reply trio/testing/_fake_net.py:27]: and indeed it did :tada: 
https://github.com/python-trio/trio/actions/runs/6567804909/job/17841165556?pr=2820#step:5:720
I just didn't see it because all runs were failing

File: trio/_core/_tests/test_multierror.py
Comment (jakkdl) [reply trio/_core/_tests/test_multierror.py:401]: should also get autofixed by ruff pyupgrade

File: trio/_tests/test_highlevel_ssl_helpers.py
Comment (jakkdl) [inline trio/_tests/test_highlevel_ssl_helpers.py:87]: @TeamSpen210 @A5rocks anybody have an idea what to do here? Looking at `SSLListener` it seems to correctly require the generic to be a `Stream`, as it explicitly says it wraps a `Stream` and calls both `self.transport_listener.send_all` and `self.transport_listener.receive_some`. So is this test doing something completely illegal/illogical solely for the sake of testing and should just be `type: ignore'd`, or is this a limitation in the API and something `SSLStream`-like should be able to wrap a `Listener`, or something else?

File: trio/testing/_fake_net.py
Comment (jakkdl) [inline trio/testing/_fake_net.py:383]: > trio/testing/_fake_net.py:(383:16 - 383:35): "UDPBinding" has no attribute "remote"  [attr-defined]

this method is never run in CI, so I suspect it's the code rather than the type hints that are wrong?

File: trio/testing/_fake_net.py
Comment (TeamSpen210) [reply trio/testing/_fake_net.py:468]: That is correct, buffers only need the `__buffer__()` method. What you can do is `memoryview(buf).nbytes`,

File: trio/_core/_tests/test_instrumentation.py
Comment (TeamSpen210) [reply trio/_core/_tests/test_instrumentation.py:34]: Yep for `contextlib.contextmanager` it needs to be a generator, since that decorator uses `.send()`/`.throw()`. The type stubs are wrong, but changing them would require too much churn...

File: trio/_tests/test_threads.py
Comment (TeamSpen210) [reply trio/_tests/test_threads.py:592]: It's a Star Trek reference indeed.

File: trio/testing/_fake_net.py
Comment (jakkdl) [reply trio/testing/_fake_net.py:468]: thanks :+1:

File: trio/testing/_fake_net.py
Comment (jakkdl) [reply trio/testing/_fake_net.py:392]: added a bunch of asserts that will hopefully elucidate anybody trying to use this about the weirdness

File: trio/testing/_fake_net.py
Comment (jakkdl) [inline trio/testing/_fake_net.py:306]: I started writing coverage tests for `_fake_net` which is how I noticed this, but leaving those for another PR.

File: trio/_tests/test_timeouts.py
Comment (CoolCat467) [inline trio/_tests/test_timeouts.py:41]: Comment does not appear to be necessary any longer

**Review by CoolCat467 — CHANGES_REQUESTED** [review]
Looks great except a few small things I noticed

**Comment by jakkdl** [comment]
hrm, pre-commit.ci is running ruff on 0.1.1 whereas check.sh is running ruff with 0.1.0 - this feels destined to bring a conflict at some point.

**Comment by CoolCat467** [comment]
> hrm, pre-commit.ci is running ruff on 0.1.1 whereas check.sh is running ruff with 0.1.0 - this feels destined to bring a conflict at some point.

I quite agree, we should update that soon.

**Review by CoolCat467 — APPROVED** [review]
Looks great, I'm sure the collections.abc things will get fixed at a later point if/when we run pyupgrade.

**Comment by CoolCat467** [comment]
In reference to commit 6170384, if you look at the pre-commit run in the PR checks area, at the very bottom of the page it displays, it says
>autofix pushing was disabled in the configuration. to autofix this pr, comment "pre-commit.ci autofix" on the pull request

**Comment by jakkdl** [comment]
I'll merge this soonish unless anybody has something to comment on, because this touches so much it very easily get merge conflicts. (e.g. other people adding test cases that aren't typechecked without this PR)

File: trio/_core/_tests/test_instrumentation.py
Comment (A5rocks) [inline trio/_core/_tests/test_instrumentation.py:15]: I think you can keep the variable length tuples and just use tuple length to differentiate (I believe that narrows in mypy? If not, ignore this comment!)

Specifically `list[tuple[str] | tuple[str, Task]]`

File: trio/_core/_tests/test_instrumentation.py
Comment (A5rocks) [inline trio/_core/_tests/test_instrumentation.py:222]: I'm not sure why this is type erroring?

File: trio/_core/_tests/test_ki.py
Comment (A5rocks) [inline trio/_core/_tests/test_ki.py:151]: Just checking understanding, this is cause `async_generator` isn't typed?

File: trio/_core/_tests/test_ki.py
Comment (A5rocks) [inline trio/_core/_tests/test_ki.py:469]: TBH we should experiment with the mypy option that allows for reassignment...

File: trio/_core/_tests/test_ki.py
Comment (A5rocks) [inline trio/_core/_tests/test_ki.py:492]: Probably more something to bring up with typeshed? But it's not documented AFAICT so probably intentional.

File: trio/_core/_tests/test_multierror.py
Comment (A5rocks) [inline trio/_core/_tests/test_multierror.py:72]: Again the comment about `-> BaseException | None` -- if mypy requires you explicitly return `None` then nvm on that.

File: trio/_core/_tests/tutil.py
Comment (A5rocks) [inline trio/_core/_tests/tutil.py:103]: Hm. I'm not sure I quite follow why the `T` here -- but I don't remember the variance of `Iterable` so maybe this makes sense... Though I guess maybe for `set`.... aaa whatever, might as well ask.

File: trio/_highlevel_ssl_helpers.py
Comment (A5rocks) [reply trio/_highlevel_ssl_helpers.py:43]: Does this passing bytes error at runtime?

File: trio/_subprocess_platform/waitid.py
Comment (A5rocks) [inline trio/_subprocess_platform/waitid.py:13]: Mind leaving a comment about why this function specifically is up here?

**Review by A5rocks — APPROVED** [review]
Well, I skipped all the test files after the actual code changes because... I mean they're just tests. Comments but none of them block anything.

File: trio/_core/_tests/test_instrumentation.py
Comment (TeamSpen210) [reply trio/_core/_tests/test_instrumentation.py:15]: It does as of last week - https://github.com/python/mypy/pull/16237. So not quite yet.

File: trio/_subprocess_platform/waitid.py
Comment (TeamSpen210) [reply trio/_subprocess_platform/waitid.py:13]: I think it's to provide a definition if imported on other platforms.

**Review by TeamSpen210 — APPROVED** [review]

File: trio/_core/_tests/test_ki.py
Comment (CoolCat467) [reply trio/_core/_tests/test_ki.py:469]: The reassignment option can open the door to lots of issues though. I will say it is convenient in a lot of cases to simply reassign something, and it can become a bit of type gymnastics with especially complex code, but I think that in the end leaving it off might be the best option. I feel like a lot of subtle issues might get suppressed with the reassignment option enabled.

**Comment by CoolCat467** [comment]
pre-commit.ci autofix

File: trio/_highlevel_ssl_helpers.py
Comment (jakkdl) [reply trio/_highlevel_ssl_helpers.py:43]: strange, can't repro any problem when reverting this edit.

File: trio/_core/_tests/test_guest_mode.py
Comment (jakkdl) [reply trio/_core/_tests/test_guest_mode.py:486]: `-> str|None` would still give "missing return statement", if you have an explicit return somewhere in the function it wants explicit returns in all branches.

File: trio/_core/_tests/test_instrumentation.py
Comment (jakkdl) [reply trio/_core/_tests/test_instrumentation.py:15]: I'll leave it as is then.

File: trio/_core/_tests/test_instrumentation.py
Comment (jakkdl) [reply trio/_core/_tests/test_instrumentation.py:222]:
```
trio/_core/_tests/test_instrumentation.py:222: error: Cannot assign to a method  [method-assign]
trio/_core/_tests/test_instrumentation.py:222: error: Incompatible types in assignment (expression has type "Callable[[Task], None]", variable has type "Callable[[Arg(Task, 'task')], None]")  [assignment]
```

If I edit `_abc.Instrument.before_task_step` to make `task` positional-only then I only get the first error. Seems like `ignore[assignment]` covers both, but made it extra explicit by ignoring both, and added comment that explains the errors.

File: trio/_core/_tests/test_ki.py
Comment (jakkdl) [reply trio/_core/_tests/test_ki.py:151]: yeah, clarifying comment

File: trio/_core/_tests/test_ki.py
Comment (jakkdl) [reply trio/_core/_tests/test_ki.py:492]: Hmm, it's only `_active` and not `__active`, so I'm pretty sure a lot of code actually inspects it or modifies it in a subclass, but yeah makes sense that typeshed sticks to the documented parts.

File: trio/_core/_tests/tutil.py
Comment (jakkdl) [reply trio/_core/_tests/tutil.py:103]: I.. don't quite follow why the signature would be confusing? Though the comment above reversing them might be the cause of confusion.
You pass it a sequence of some thing, e.g. `[1, 2, 3, 4, 5]` or `[1, 2, 3, 5, 4]`, and use this function to see whether that matches a template - which can include either objects of the same type, or if you expect the order to sometimes be jumbled you can specify them inside a set*: `[1, 2, 3, {4, 5}]`. So the template should be an `Iterable` which contains objects of the same type your sequence consists of, or sets of that object. Variance is not related in any way.

The `Sequence` vs `Iterable` is because the function does slicing on `seq`, but only iterates over `template`.

\* I did not expect this to be how the function actually worked, and wrote an incorrect description at first where I assumed the sets were about unions. Not a great name imho.

Improved the comments accompanying the function.

File: trio/_subprocess_platform/waitid.py
Comment (jakkdl) [reply trio/_subprocess_platform/waitid.py:13]: It's for `trio/_tests/test_subprocess.py:test_waitid_eintr()`:
```python
@slow
def test_waitid_eintr() -> None:
    # This only matters on PyPy (where we're coding EINTR handling
    # ourselves) but the test works on all waitid platforms.
    from .._subprocess_platform import wait_child_exiting

    if TYPE_CHECKING and sys.platform == "win32":
        return

    if not wait_child_exiting.__module__.endswith("waitid"):
        pytest.skip("waitid only")
    from .._subprocess_platform.waitid import sync_wait_reapable

    ...
```

where despite having a
`if TYPE_CHECKING and sys.platform == "win32": return` it will still complain about the import on line 522. I guess this one could be reported upstream?

Though I realize now it's probably better to silence the error in the test than to make people think they can import `sync_wait_reapable` globally on all platforms only to get an `ImportError` at runtime.

File: trio/_tests/test_exports.py
Comment (jakkdl) [reply trio/_tests/test_exports.py:449]: nope. moving it up.

File: trio/_tests/test_highlevel_serve_listeners.py
Comment (jakkdl) [reply trio/_tests/test_highlevel_serve_listeners.py:24]: Also found two more places where it could be used in the file.

File: trio/_tests/test_ssl.py
Comment (TeamSpen210) [inline trio/_tests/test_ssl.py:118]: Putting these into a tuple only to immediately unpack seems a little redundant.

Comment (jakkdl) [reply trio/_tests/test_ssl.py]: indeed

File: trio/_core/_tests/tutil.py
Comment (A5rocks) [reply trio/_core/_tests/tutil.py:105]: Well the `T` w/ `Sequence` (covariant) means that something like this would pass typechecking: `check_sequence_matches([5.0, 4.0], [5, 4])` but I was in fact being blind to the fact that `object` has that problem too.

Comment (jakkdl) [reply trio/_core/_tests/tutil.py:105]: Ah, huh. Yeah I don't think we really care about that level of precision, as it's only really used in a few tests and not a public function. I need to get better at variance, for sure.

**Pull request merged**
